2-Draft-Accepted/RFCxxxx-DSC-for-PowerShell7.md

```diff
@@ -0,0 +1,323 @@
+are used to parse information store in Managed Object Framework (MOF) file format,
+which is used for both configuration files and schema files.
+
+### Challenges with the current versions of DSC
+
+Very few community authored DSC resources exist for Linux because C++ is the only
+supported language. Other than using the Script/nxScript resources combined with
+a composite resource, it is near-impossible today to create a DSC solution that is cross-platform.
+
+Today, third party solutions leveraging DSC resources need to call the LCM
+to run Get/Set/Test, making it complicated to integrate with partner ecosystems.
+
+## Scope of DSC implementation for PowerShell 7
+
+Unlike DSC for Windows PowerShell, DSC for PowerShell 7 is only intended to provide
+a platform for building or integrating with configuration solutions. For PowerShell 7,
+the scope of DSC support is limited to the `configuration` keyword, authoring,
+compilation, and using DSC resources via the PSDesiredStateConfiguration module.
+
+This means that a default agent (LCM) is not included and existing pull servers are not supported.
+
+### Motivation
+
+DSC in PowerShell 7 should enable the following user scenarios:
+
+    As a developer of configuration management solutions,
+    I can easily integrate DSC resources
+    so that my solution can leverage the PowerShell ecosystem and its users and contributors.
+
+    As an IT Pro,
+    I can directly invoke DSC resources
+    so that I can build ad hoc cross platform configurations.
+
+### Cross platform first
+
+DSC has supported both Windows and Linux, however, the
+[Linux support](https://docs.microsoft.com/powershell/scripting/dsc/getting-started/lnxGettingStarted)
+was dependent on a Microsoft-maintained implementation of
+[OMI](https://github.com/Microsoft/omi)
+and did not encourage authoring Linux DSC resources in PowerShell,
+making development inconsistent and difficult for non-expert users.
+
+DSC for PowerShell 7 will prioritize being cross platform and having a consistent
+user experience and capabilities on all supported platforms.
+
+### Invoke-DscResource
+
+About a year ago, we worked with
+[Gael Colas](https://twitter.com/gaelcolas)
+to author an RFC describing fundamental changes to
+[`Invoke-DscResource`](https://docs.microsoft.com/powershell/module/psdesiredstateconfiguration/invoke-dscresource)
+in PowerShell 7.0.
+This enabled executing the Get, Set, and Test methods on DSC resources bypassing
+the LCM on Windows and removing the need for OMI on Linux. This was the start
+to enable cross platform DSC resources to be authored in PowerShell script (with
+some limitations).
+
+The changes to DSC shipped as an
+[experimental feature](https://docs.microsoft.com/powershell/scripting/learn/experimental-features)
+in PowerShell 7.0.
+
+### Bring your own agent (BYOA)
+
+A significant change from existing DSC for Windows PowerShell is there will be no
```

Comment:
I'm still not clear on what I need to do here. Do we have documentation to which we can refer for enabling this.

Reply:
This is confusing to me as well.  If I'm understanding correctly this would be the new workflow:

1. Write out a configuration in PowerShell (using the Configure keyword), same as previous versions.
2. Export the configuration to JSON.
3. Use a custom script or third-party tool to parse the JSON and then run the DSC resource modules using Invoke-DscResource.

I think step 3 is where my disconnect is.  Could we have a built in command that would would parse the JSON file and run the DSC resources?  This would at least give us the base functionality, which we could then extend if needed.

Reply:
since there is no schema, I think it wont be an easy thig to provide a built-in cmdlet for that and I believe the idea is to enable anyone to create an agent which may or may not be in PowerShell.

Reply:
Well there doesn’t need to be a strict schema. The configuration file is basically just key value pairs. All DSC resources have get, test, and set methods you use to compare with the values in the config.  That’s what the original LCM did. Granted it wouldn’t be very hard to write a script to reproduce that functionality, but I feel like by not having this built in, it will prevent a lot of non-programmers from using DSC. I mean the original point of DSC was so non-programmers could easily build up configurations. If you really strip things down to basics, we don’t even really need DSC resources. We could just write PowerShell scripts that apply the settings we want directly.

Reply:
yep, I agree

Reply:
A small clarification about ...

> If I'm understanding correctly this would be the new workflow:
> Write out a configuration in PowerShell (using the Configure keyword), same as previous versions.
> Export the configuration to JSON.
> Use a custom script or third-party tool to parse the JSON and then run the DSC resource modules using Invoke-DscResource.

This is one option; that is if the user decides to go through `"configuration ps1" -> JSON (which used to be MOF) -> third-party tool to parse the JSON and then do whatever it needs` route.

Another (faster) option is to [just call `Invoke-DscResource` directly - see examples in this link](https://docs.microsoft.com/en-us/powershell/module/psdesiredstateconfiguration/invoke-dscresource?view=powershell-7).

Reply:
I understand. My main point was that one of the original benefits of DSC was that people without a lot of programming or scripting experience could still write configurations. Having to manually call the resources with Invoke-DscResource, or build a script to replace the functionality of the LCM, just seems like a bit of a step back. I like the idea of having it open for third-parties to build on, but it feels like we are loosing some base functionality in the process. Anyway, that’s just my 2 cents.

Reply:
I kinda agree with you, but we lost that already from 5.1.
This is actually "bringing back" functionality, so I see it as a good (re)start.
I hope that in due time we'll get closer to what we can do with 5.1, whether it's with a third party agent, from the community or why not using Guest Config...

Reply:
Hey,

I like the discussion so far. But I must agree with @kgraves83 a lot. By removing the LCM completely, it will be hard for people to manage DSC. I know in the RFC it's talking about: ` integrate with existing configuration agents like Ansible, Chef, Puppet, Azure Policy Guest Configuration, and other popular configuration solutions.` But I don't think that is a good solution for companies, who use the `old` DSC. Because you can just run `Start-DSCConfiguration` on any machine in a well-configured environment and the party can start.

With the new solution, you must first find the current `json` configurations for the machine. Then you must take a look into the configuration and install all required modules on the target machine. (normally DSC would pull them from the pullserver) And then run them on each target machine. And the tool to manage all of this is handcrafted by an admin. Oh and don't forget the monitoring feature. That must be also done manually with the new solution if I understand it right.

I mean it's a good starting point to implement first the logic of "how a configuration is parsed and can be used". But this RFC shouldn't forget the management of many configurations for a big DSC environment. The popular configuration solutions had for me two downsides:

- Ansibel and Chef needs a Linux Host
- Azure Policy Guest Configuration works only if the company uses azure

So in a completely Windows based environment that is disconnected, there is no solution. I really want to see something like a basic tooling for those use cases. I think the idea of something like a LCM as agent isn't a bad idea. I think DSC need something like that to work well. The downsides of not having it are to big. Maybe this is the wrong RFC for that discussion , but I think that is important.

I really wish to see something like LCM 2.0 or something like that.

Reply:
Worth noting here: From a CM vendor perspective (`/me waves in Puppet developer`) while we _could_ support calling a config script from within the CM tool, they already have a DSL (or yaml def in the case of Ansible) where users are defining their full configuration info - even telling DSC to handle the JSON blob of configs would require writing some DSL - we're probably going to continue to support and recommend our integration which is built on top of `Invoke-DscResource`.

For Puppet specifically, we've [been doing a lot of work over the last](https://github.com/puppetlabs/Puppet.Dsc) year to build a strong translation layer between the API surface of DSC Resources and Puppet so you can consume DSC Resources _just like_ any other Puppet resource - with the full suite of support sugar from our VSCode extension, compile time validation, etc. 

I think that experience is going to be better for our users than a new Puppet resource which allows you to specify a JSON blob (as a herestring or path to the blob) and is only able to do change reporting based on the entire blob or (assuming property-by-property reporting gets implemented at the same time as these other changes) needs custom logic to surface DSC's internal change reporting.

I _suspect_ but do not _know_ that the situation for other CM vendors is similar from an integration perspective.

The more important need for us is to make sure we can fully map a DSC Resource's surface from `Get-DscResource` since we won't be able to map it via inspecting the CIM instances anymore (especially whatever we're calling nested CIM instances now that they won't be CIM instances).